applications/ShapeOptimizationApplication/python_scripts/custom_math.py

```diff
@@ -0,0 +1,422 @@
+# ==============================================================================
```

Comment:
I don't know the details why you need this, but **reimplementing** linear algebra **in python** does not seem like a good idea...
This might work for small problems but as soon as you have sth bigger I suspect it to be slow

I guess you do this partially bcs you cannot use numpy?

=> Do you know that you can use the Kratos-Vectors and matrices also in Python? Both Dense (some basic examples [here](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/tests/test_vector_interface.py)) and Sparse?
If functionalities are missing you could consider implementing them in the core such that also others can use it

Not everything might be supported but those things you could implement yourself if you have to
Then you could also use e.g. the linear solvers...